File: Week03/17.电话号码的字母组合.py
#
# @lc app=leetcode.cn id=17 lang=python3
#
# [17] 电话号码的字母组合
#

# @lc code=start
class Solution:
    def letterCombinations1(self, digits: str) -> List[str]:
        # time: 32ms, beats 94.95%
        # define a dict
        num_letter_dict = {'2': 'abc', '3': 'def',  
                            '4': 'ghi', '5': 'jkl',
                            '6': 'mno', '7': 'pqrs',  
                            '8': 'tuv', '9': 'wxyz' }
        if len(digits) == 0: 
            return []
        if len(digits) == 1:
            return list(num_letter_dict[digits[0]])
        # similar to "ziji": previous + the last one
        prev = self.letterCombinations(digits[:-1])
        additional = num_letter_dict[digits[-1]]
        return [s + c for s in prev for c in additional]


    def letterCombinations(self, digits: str) -> List[str]:     
    # backtracking
    # 52ms, beats 12.36%
    if not digits:
        return []
    num_letter_dict = {'2': 'abc', '3': 'def',  
                '4': 'ghi', '5': 'jkl',
                '6': 'mno', '7': 'pqrs',  
                '8': 'tuv', '9': 'wxyz' }
    res = []
    self.helper(digits, num_letter_dict, 0, "", res)
    return res

    def helper(self, digits, num_letter_dict, idx, path, res):
    if len(path) == len(digits):
        res.append(path)
        return

    for i in range(idx, len(digits)):
        for j in num_letter_dict[digits[i]]:
            self.helper(digits, num_letter_dict, i + 1, path + j, res)


# @lc code=end

File: Week03/17.电话号码的字母组合.java
/*
 * @lc app=leetcode.cn id=17 lang=java
 *
 * [17] 电话号码的字母组合
 */

// @lc code=start
class Solution {
    public List<String> letterCombinations(String digits) {
        if (digits == null || digits.length() == 0) {
            return new ArrayList();
        }
        Map<Character, String> map = new HashMap<Character, String>();
        map.put('2', "abc");
        map.put('3', "def");
        map.put('4', "ghi");
        map.put('5', "jkl");
        map.put('6', "mno");
        map.put('7', "pqrs");
        map.put('8', "tuv");
        map.put('9', "wxyz");
        List<String> res = new LinkedList<String>();
        search("", digits, 0, res, map);
        return res;
    }

    private void search(String s, 
                        String digits, 
                        int i,
                        List<String> res,
                        Map<Character, String> map) {
        // terminator
        if (i == digits.length()) {
            res.add(s);
            return;
        }

        // process
        String letters = map.get(digits.charAt(i));
        for (int j = 0; j < letters.length(); j++) {
            // drill down
            search(s + letters.charAt(j), digits, i+1, res, map);         
        }

        // reverse
    }
}
// @lc code=end

